control-plane-operator/controllers/hostedcontrolplane/cco/reconcile.go

```diff
@@ -0,0 +1,486 @@
+	apiPort := config.DefaultAPIServerPort
+	if params.apiPort != nil {
+		apiPort = int(*params.apiPort)
+	}
+
+	params.OwnerRef.ApplyTo(deployment)
+	deployment.Spec = appsv1.DeploymentSpec{
+		Selector: &metav1.LabelSelector{
+			MatchLabels: selectorLabels(),
+		},
+		Strategy: appsv1.DeploymentStrategy{
+			Type: appsv1.RecreateDeploymentStrategyType,
+		},
+		Template: corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: selectorLabels(),
+			},
+			Spec: corev1.PodSpec{
+				AutomountServiceAccountToken: pointer.Bool(false),
+				Containers: []corev1.Container{
+					util.BuildContainer(containerMain(), buildMainContainer(params.operatorImage, params.releaseVersion, apiPort)),
+					util.BuildContainer(containerMetrics(), buildMetricsContainer(params.kubeRbacProxyImage, apiPort)),
+					util.BuildContainer(containerClientTokenMinter(), buildClientTokenMinter(params.tokenMinterImage, params.issuerURL)),
+				},
+				Volumes: []corev1.Volume{
+					util.BuildVolume(volumeClientToken(), buildVolumeClientToken),
+					util.BuildVolume(volumeServingCert(), buildVolumeServingCert),
+					util.BuildVolume(volumeAdminKubeconfig(), buildVolumeAdminKubeconfig),
+					util.BuildVolume(volumeCABundle(), buildVolumeCABundle),
+				},
+			},
+		},
+	}
+
+	params.deploymentConfig.ApplyTo(deployment)
+	util.AvailabilityProber(kas.InClusterKASReadyURL(deployment.Namespace, params.apiPort), params.availabilityProberImage, &deployment.Spec.Template.Spec, func(o *util.AvailabilityProberOpts) {
+		o.KubeconfigVolumeName = "kubeconfig"
+		o.WaitForInfrastructureResource = true
+		o.RequiredAPIs = []schema.GroupVersionKind{
+			{Group: "operator.openshift.io", Version: "v1", Kind: "CloudCredential"},
+		}
+	})
+	return nil
+}
+
+func containerMain() *corev1.Container {
+	return &corev1.Container{
+		Name: "cloud-credential-operator",
+	}
+}
+
+func renderOperatorStartScript(startScriptTemplate *template.Template) (string, error) {
+	out := &bytes.Buffer{}
+	params := struct {
+		TokenDir        string
+		TokenFile       string
+		WorkerNamespace string
+		CABundle        string
+	}{
+		TokenDir:        volumeMounts.Path(containerMain().Name, volumeClientToken().Name),
+		TokenFile:       kubeconfigFile,
+		WorkerNamespace: workerNamespace,
+		CABundle:        path.Join(volumeMounts.Path(containerMain().Name, volumeCABundle().Name), certs.CASignerCertMapKey),
+	}
+	if err := startScriptTemplate.Execute(out, params); err != nil {
+		return "", err
+	}
+	return out.String(), nil
+}
+
+func buildMainContainer(image, releaseVersion string, apiPort int) func(*corev1.Container) {
+	return func(c *corev1.Container) {
+		c.Image = image
+		c.Command = []string{
+			"/bin/bash",
+		}
+		c.Args = []string{
+			"-ec",
+			startScript,
+		}
+		c.Env = []corev1.EnvVar{
+			{
+				Name:  "RELEASE_VERSION",
+				Value: releaseVersion,
+			},
+			{
+				Name:  "KUBERNETES_SERVICE_HOST",
```

Comment:
nit: these env vars and volume(mounts) are also things that could be injected as needed by a `token-minter` utility